benchmarks/profiler_heap_serialization.rb

```diff
@@ -0,0 +1,249 @@
+
+def sample_object(recorder, obj, depth = 0)
+  if depth <= 0
+    Datadog::Profiling::StackRecorder::Testing._native_track_object(
+      recorder,
+      obj,
+      1,
+      obj.class.name,
+    )
+    Datadog::Profiling::Collectors::Stack::Testing._native_sample(
+      Thread.current,
+      recorder,
+      METRIC_VALUES,
+      [],
+      [],
+      400,
+      false
+    )
+  else
+    sample_object(recorder, obj, depth - 1)
+  end
+end
+
+def random_value(depth = 0)
+  case [:array, :hash, :string, :number, :object].sample
+  when :array
+    random_array(depth)
+  when :hash
+    random_hash(depth)
+  when :string
+    RANDOM.bytes(rand(0..50))
+  when :number
+    rand(-1_000_000_000_000..1_000_000_000_000)
+  when :object
+    BenchmarkObject.new(depth)
+  end
+end
+
+def create_benchmark_live_objects(num_objects: 1_000_000)
+  live_objects = []
+
+  num_objects.times { live_objects << random_object }
+end
+
+def random_array(depth)
+  num_entries = rand(0..(100 / (10**depth)))
+  Array.new(num_entries) { random_value(depth + 1) }
+end
+
+def random_hash(depth)
+  num_entries = rand(0..(100 / (10**depth)))
+  Array.new(num_entries) { [random_value(depth + 1).freeze, random_value(depth + 1)] }.to_h
+end
+
+class BenchmarkObject < Object
+  def initialize(depth)
+    num_fields = rand(0..(100 / (10**depth)))
+    obj = self
+
+    num_fields.times { |i| obj.instance_variable_set("@field#{i}".to_sym, random_value(depth + 1)) }
+  end
+end
+
+# This benchmark measures the performance of different flavours of heap profiling
+Benchmark.ips do |x|
+  benchmark_time = VALIDATE_BENCHMARK_MODE ? { time: 0.01, warmup: 0 } : { time: 10, warmup: 2 }
+  x.config(
+    **benchmark_time,
+    suite: report_to_dogstatsd_if_enabled_via_environment_variable(benchmark_name: 'profiler_heap_serialization')
+  )
+
+  recorder_without_heap = Datadog::Profiling::StackRecorder.new(
+    cpu_time_enabled: false,
+    alloc_samples_enabled: false,
+    heap_samples_enabled: false,
+    heap_size_enabled: false,
+    heap_sample_every: 1,
+    timeline_enabled: false,
+  )
+
+  recorder_with_heap_samples_only = Datadog::Profiling::StackRecorder.new(
+    cpu_time_enabled: false,
+    alloc_samples_enabled: true,
+    heap_samples_enabled: true,
+    heap_size_enabled: false,
+    heap_sample_every: 1,
+    timeline_enabled: false,
+  )
+
+  recorder_with_full_heap = Datadog::Profiling::StackRecorder.new(
+    cpu_time_enabled: false,
+    alloc_samples_enabled: true,
+    heap_samples_enabled: true,
+    heap_size_enabled: true,
+    heap_sample_every: 1,
+    timeline_enabled: false,
+  )
+
+  live_objects = Array.new(VALIDATE_BENCHMARK_MODE ? 100 : 10_000) { BenchmarkObject.new(0) }
+
+  live_objects.each_with_index do |obj, _i|
+    sample_object(recorder_without_heap, obj)
+    sample_object(recorder_with_heap_samples_only, obj)
+    sample_object(recorder_with_full_heap, obj)
+  end
```

Comment:
Wait, I'm somewhat confused by what's going on here. Each `BenchmarkObject` holds inside it references to a bunch of objects that are the result of calling `random_value`.

But... since we only really sample the `BenchmarkObject` instance itself... how does all that fancyness differ from sampling `Array.new(...) { Object.new }`? As far as the profiler goes, it never sees anything other that instances of `BenchmarkObject`, so all of the work above to randomly do stuff seems mostly a no-op? Am I missing something O_o...?